datasapiens/qa-webapp
# [Ready-To-Review] Added automated scripts for Income Expense App

**Summary/Motivation:**
The PR contains a complete automation framework designed using [Cypress](cypress.io) which automates a few test cases for the Income Expense app. 

**Documentation:**
Please refer to the **README.md** file inside e2e folder for added documentation of the automation script. 

**Test Cases Covered:**
- [x] Verify sign-up of multiple users.
- [x] Verify log-in functionality.
- [x] Verify budgets can be created.
- [x] Verify budget can be allocated to another user.
- [x] Verify multiple categories can be added to a single user budget.
- [x] Verify multiple categories can be added to a multi-user budget.
- [x] Compare the expense for categories with respect to users in Stats page.
- [x] Verify deletion of the budget created.

**Dependencies:**
- In order to use test runner locally, node and cypress must be installed. To install Cypress - `npm install cypress --save-dev`.

**Screenshots, Videos or GIF:**
Please find the automated test run recorded from the test runner: 

https://user-images.githubusercontent.com/41227164/156840028-ef4c396c-8742-435d-9c79-ed6eb23c36f6.mov

**Deployment/Testing Strategy:**
Multiple deployment/testing strategies have been added for easy execution of the test scripts. Below are the details regarding these strategies.
1. Using shell file
- The shell file can be used to run all services and automation suite using a single command.

````
./cypress-run.sh
````

2. Using Makefile
- A makefile has been included to perform some of the basic tasks required.

- Build services: 
```
make docker-build
````
- Run services: 
```
make docker-up
````
- Run tests headlessly: 
```
make test
````
- Run tests using a test runner:
```
make test-gui
```

3. Using Cypress Test Runner
- Test runner can be built locally and used to test the automated scripts. For this test runner should be opened. In order to open the test run, go inside the `e2e` folder and then use the command below:
```
npx cypress open
```
